openedx/features/discounts/utils.py

```diff
@@ -50,52 +50,110 @@ def offer_banner_wrapper(user, block, view, frag, context):  # pylint: disable=W
+        discounted_price = base_price * ((100.0 - percentage) / 100)
+        if discounted_price == int(discounted_price):
+            discounted_price = "{:0.0f}".format(discounted_price)
         else:
-            discount_price = format_course_price("{:0.2f}".format(discount_price))
+            discounted_price = "{:0.2f}".format(discounted_price)
+
+        return format_course_price(base_price), format_course_price(discounted_price), percentage
+    else:
+        return format_course_price(base_price), None, None
+
+
+def generate_offer_data(user, course):
+    """
+    Create a dictionary of information about the current discount offer.
+
+    Used by serializers to pass onto frontends and by the LMS locally to generate HTML for template rendering.
+
+    Returns a dictionary of data, or None if no offer is applicable.
+    """
+    if not user or not course or user.is_anonymous:
+        return None
+
+    ExperimentData.objects.get_or_create(
+        user=user, experiment_id=REV1008_EXPERIMENT_ID, key=str(course),
+        defaults={
+            'value': datetime.now(tz=pytz.UTC).strftime('%Y-%m-%d %H:%M:%S%z'),
+        },
+    )
+
+    expiration_date = get_discount_expiration_date(user, course)
+    if not expiration_date:
+        return None
 
-        # Separate out this string because it has a lot of syntax but no actual information for
-        # translators to translate
-        formatted_discount_price = HTML(
-            u"{s_dp}{discount_price}{e_p} {s_st}{s_op}{original_price}{e_p}{e_st}"
-        ).format(
+    if not can_receive_discount(user=user, course=course, discount_expiration_date=expiration_date):
+        return None
+
+    original, discounted, percentage = _get_discount_prices(user, course, assume_discount=True)
+
+    return {
+        'code': 'BIENVENIDOAEDX' if get_language() == 'es-419' else 'EDXWELCOME',
+        'expiration_date': expiration_date,
+        'original_price': original,
+        'discounted_price': discounted,
+        'percentage': percentage,
+        'upgrade_url': verified_upgrade_deadline_link(user, course=course),
+    }
+
+
+def _format_discounted_price(original_price, discount_price):
+    """Helper method that returns HTML containing a strikeout price with discount."""
+    # Separate out this string because it has a lot of syntax but no actual information for
+    # translators to translate
+    formatted_discount_price = HTML(
+        u"{s_dp}{discount_price}{e_p} {s_st}{s_op}{original_price}{e_p}{e_st}"
+    ).format(
+        original_price=original_price,
+        discount_price=discount_price,
+        s_op=HTML("<span class='price original'>"),
+        s_dp=HTML("<span class='price discount'>"),
+        s_st=HTML("<del aria-hidden='true'>"),
+        e_p=HTML("</span>"),
+        e_st=HTML("</del>"),
+    )
+
+    return (
+        HTML(_(
+            u"{s_sr}Original price: {s_op}{original_price}{e_p}, discount price: {e_sr}{formatted_discount_price}"
```

Comment:
Nit: remove the `u`'s in these functions.

Super optional nit: Switch from `"` to `'`. I can't tell what's the dominant one in this file, but I see both being used and I know we tend to prefer single quotes so I'd opt for switching to those.

Reply:
Good catch. Yeah pylint prefer single quotes (unless you have a single quote already inside the string, then it lets you use double quotes to avoid a backslash).